### PR TITLE
[Merged by Bors] - feat: port Algebra.Order.Monoid.Canonical.Defs

### PR DESCRIPTION
**Pull request by rosborn**
mathlib SHA: 1fc36cc9c8264e6e81253f88be7fb2cb6c92d76a

Porting Notes:
~~1. Haven't touched `protect_proj`. If someone knows what to do to fix this, please feel free.~~
~~2. `LinearOrder.toLattice` is found in `Order.Lattice` which is both not ported yet and not imported.~~
3. `simp_rw` seems to be more aggressive in lean4, so a proof was rewritten to just use `rw` alone.
4. `simpa` seemed to continue to try and solve a goal after it had already been solved.
5. `one_lt_mul_iff` needed an additional `rfl` to solve. This wasn't needed in lean3.
6. Some `simp` lemmas are not simp-normal form. I have not debugged any of these yet.
7. Further documentation is needed. If anyone wants to add them, please feel free.

### CODE REVIEW

**Comment by semorrison** [comment]
1. Just remove `protect_proj`, and write `protected` in front of each field name.

**Comment by semorrison** [comment]
4. Try replacing `simpa using f` with `simp at f`. If `simp` can reduce it to `False`, then it will close the goal, thus making `simpa` fail because it then wants to run `assumption`. This is a change in behaviour of tactics that we probably should add a warning for.

**Comment by semorrison** [comment]
2. `Order.Lattice` has been ported? Could you double check on this. It should be imported via Order.MinMax.

File: Mathlib/Algebra/Order/Monoid/Canonical/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+theorem le_mul_right (h : a ≤ b) : a ≤ b * c :=
+  calc
+    a = a * 1 := by simp
+    _ ≤ b * c := mul_le_mul' h (one_le _)
+
+#align le_mul_right le_mul_right
+
+@[to_additive]
+theorem lt_iff_exists_mul [CovariantClass α α (· * ·) (· < ·)] : a < b ↔ ∃ c > 1, b = a * c := by
+  -- Porting note: simp_rw was more aggressive in lean4
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Monoid/Canonical/Defs.lean:251]: In cases like this, could you leave the original proof in place, commented out? This makes it easier for someone to address the problem, without having to dig up other repositories.

File: Mathlib/Algebra/Order/Monoid/Canonical/Defs.lean [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+#align canonically_ordered_add_monoid CanonicallyOrderedAddMonoid
+
+
+-- see Note [lower instance priority]
+instance (priority := 100) CanonicallyOrderedAddMonoid.toOrderBot (α : Type u)
+    [h : CanonicallyOrderedAddMonoid α] : OrderBot α :=
+  { h with }
+#align canonically_ordered_add_monoid.to_order_bot CanonicallyOrderedAddMonoid.toOrderBot
+
+-- Porting note: Should this documentation string differs from `CanonicallyOrderedAddMonoid`
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Monoid/Canonical/Defs.lean:107]: Yes, this is fine. The extra comments are useful.
```suggestion
```

**Comment by semorrison** [comment]
Actually looks pretty close, otherwise!

**Comment by rosborn** [comment]
> 1. Just remove `protect_proj`, and write `protected` in front of each field name.

The definitions for `CanonicallyLinearOrderedAddMonoid` and `CanonicallyLinearOrderedMonoid` only extend other classes, so there are no additional fields. Was `protect_proj` doing anything for those classes? Is it safe to remove the attribute without any other changes?

**Comment by rosborn** [comment]
Should the `NeZero` instance for `bit0` be deleted?

**Comment by semorrison** [comment]
> Should the `NeZero` instance for `bit0` be deleted?

No, we're leaving stuff about `bit0` in place for now. We'll need to turn off the deprecated warning.

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge

**Comment by riccardobrasca** [comment]
bors merge

**Pull request closed without merging**